### PR TITLE
feat(customs): limit avatar uploads

### LINKED ISSUE #132: Rate-limit avatar upload attempts using fxa-customs-server?
Avatar uploads are a potentially expensive operation that we could profitably rate-limit by IP-address and/or email-address.  It may be possible to use fxa-customs-server for this without too much work.  @seanmonstar what do you think about the potential benefits vs potential implementation costs of doing so?

### PR DESCRIPTION
**Pull request by larissagaulia**
Fixes #132

### CODE REVIEW

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: f52c89b60feb569f022533d251e01760eb6ef97f
  * Commits must be in the following format: **%{type}(%{scope}): %{description}**

 * Commit: 95d5bcfcdf66a5876d017a6f588b747b4e7bda75
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-profile-server

---
This message was auto-generated by https://gitcop.com

File: lib/pool.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  return this.request('GET', path)
+}
+
+Pool.prototype.del = function (path, data) {
+  return this.request('DELETE', path, data)
+}
+
+Pool.prototype.head = function (path) {
+  return this.request('HEAD', path)
+}
```
Comment (vladikoff) [inline lib/pool.js]: @larissagaulia from the above I think you only need `.post`

File: lib/customs.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var P = require('./promise')
+var Pool = require('./pool')
+
+function Customs() {
+  this.pool = new Pool("http://127.0.0.1:7000", { timeout: 1000 })
```
Comment (vladikoff) [inline lib/customs.js]: need this one to use config. see how we `require` config in other modules and pull the customs url from the config.

File: lib/routes/avatar/post.js [hunk 3]
```diff
@@ -54,15 +57,19 @@ module.exports = {
...
-    .done(function() {
-      var info = {
-        event: 'avatar.post',
-        uid: uid
-      };
-      logger.info('activityEvent', info);
-      reply({ id: hex(id) }).code(201);
-    }, reply);
+
+    customs.checkAuthenticated('127.0.0.1', token, uid)
```
Comment (vladikoff) [inline lib/routes/avatar/post.js]: Add the proper  ip address by adding: https://github.com/mozilla/fxa-auth-server/blob/10ee3224f271f48500b84c96786ce78c95c5a7f6/lib/server.js#L217 to fxa-profile-server server.js and use `request.app.clientAddress` (example https://github.com/mozilla/fxa-auth-server/blob/master/lib/routes/account.js#L368)

File: lib/config.js [hunk 4]
```diff
@@ -177,6 +177,11 @@ const conf = convict({
       default: 'http://127.0.0.1:9010/v1'
     }
   },
+  customsUrl: {
+    doc: 'fraud / abuse server url',
+    default: 'http://127.0.0.1:7000',
```
Comment (vladikoff) [inline lib/config.js]: @larissagaulia this is good. to [follow the auth server configuration](https://github.com/mozilla/fxa-auth-server/blob/master/.env.dev#L2) put `none` into config override https://github.com/mozilla/fxa-profile-server/blob/master/config/dev.json . For your development use `CUSTOMS_SERVER_URL=http://127.0.0.1:7000 npm start` or override dev.json without committing it.

File: lib/customs.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+Customs.prototype.checkAuthenticated = function (ip, token, uid) {
+  return this.pool.post(
+      '/checkAuthenticated',
+      {
+        ip: ip,
+        token: token,
+        uid: uid
+      }
+  )
```
Comment (vladikoff) [inline lib/customs.js]: @larissagaulia I guess this also needs to use the `tooManyRequests` error: https://github.com/mozilla/fxa-auth-server/blob/7ca2ffba99e09a1e0025baa26bbaca35598d73bd/lib/customs.js#L51

**Comment by jrgm** [comment]
May I request that this not be merged into master, when it's ready, until later this week. I'd like to wrap up a 0.65.0 release, possibly with the `grunt version` PR and a revert PR for NODE_ENV changes. Apologies, but I'm a bear of little brain.

File: lib/customs.js (hunk 5 above)
Comment (rfk) [inline lib/customs.js]: Yes please, and also note it in the API docs for the relevant endpoints, as done at e.g. https://github.com/mozilla/fxa-auth-server/blob/master/docs/api.md#response-2

File: lib/routes/avatar/post.js (hunk 3 above)
Comment (vladikoff) [inline lib/routes/avatar/post.js]: @dannycoates @seanmonstar @rfk I forgot, do we to do the whole dance in `onPreAuth ` like here: https://github.com/mozilla/fxa-auth-server/blob/10ee3224f271f48500b84c96786ce78c95c5a7f6/lib/server.js#L217  to get a proper IP or not?

Comment (rfk) [inline lib/routes/avatar/post.js]: It's probably worth doing similar logic, yeah.

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: dbbf4957c84bef04f1f33ce85510d9a1ba680391
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

 * Commit: 5198d41c834db24ea0f545eed3a015dab7fd486c
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

 * Commit: afaa0ff35ccb9cd91ada5cf588bd726c95d4c154
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-profile-server

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 116d6c9fa63f5544a3dc635aa3afc133beb06cd6
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-profile-server

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 116d6c9fa63f5544a3dc635aa3afc133beb06cd6
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

 * Commit: 16db556e944e60f2d98d7deac9f10102d7a8c8fd
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-profile-server

---
This message was auto-generated by https://gitcop.com

File: config/dev.json [hunk 6]
```diff
@@ -6,5 +6,6 @@
   },
   "img": {
     "driver": "local"
-  }
+  },
+  "CUSTOMS_SERVER_URL": "none"
```
Comment (vladikoff) [inline config/dev.json]: this should be something like:

```
"customsUrl": "none"
```

these need to be the same as the keys in lib/config.js, not env variables.

File: docs/API.md [hunk 7]
```diff
@@ -45,6 +45,8 @@ The currently-defined error responses are:
 - 403, 100: Unauthorized
 - 400, 101: Invalid request parameter
 - 400, 102: Unsupported image provider
+- 400, 125: The request was blocked for security reasons
+- 429, 114: Client has sent too many requests
```
Comment (vladikoff) [inline docs/API.md]: 👏 nice

File: lib/customs.js [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+Customs.prototype.checkAuthenticated = function (action, ip, token, uid) {
+  logger.info('customs.checkAuthenticated', { ip: ip, token: token, uid: uid});
+
+  return this.pool.post(
+    '/checkAuthenticated',
+    {
+      action: action,
+      ip: ip,
+      token: token,
```
Comment (vladikoff) [inline lib/customs.js]: remove all mentions of `token`, we don't accept `token` in the customs server

File: lib/error.js [hunk 9]
```diff
@@ -116,6 +116,37 @@ AppError.unsupportedProvider = function unsupportedProvider(url) {
   });
 };
 
+//TODO: see about errno. Here using the same as auth server
```
Comment (vladikoff) [inline lib/error.js]: @larissagaulia same errno is okay. were there any other concerns?

File: lib/error.js [hunk 10]
```diff
@@ -116,6 +116,37 @@ AppError.unsupportedProvider = function unsupportedProvider(url) {
...
+  return new AppError({
+    code: 400,
+    error: 'Request blocked',
+    errno: 125,
+    message: 'The request was blocked for security reasons'
+  });
+};
+
+AppError.tooManyRequests = function (retryAfter) {
+
```
Comment (vladikoff) [inline lib/error.js]: remove line break

File: lib/routes/avatar/post.js [hunk 11]
```diff
@@ -40,9 +40,14 @@ module.exports = {
     }
   },
   handler: function avatarPost(req, reply) {
-    var uid = req.auth.credentials.user;
+
+    var action = 'avatarUpload';
+    var ip = req.app.clientAddress;
     var payload = req.payload;
     var provider;
+    var token = req.auth.credentials.token;
```
Comment (vladikoff) [inline lib/routes/avatar/post.js]: remove token

**Comment by vladikoff** [comment]
@larissagaulia almost ready to go 👍  see notes above

**Comment by vladikoff** [comment]
@larissagaulia also you don't have to write `WIP` in commit messages, such as `WIP feat(server)....`. adding a github label or changing the title in GitHub is fine .

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 116d6c9fa63f5544a3dc635aa3afc133beb06cd6
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

 * Commit: 16db556e944e60f2d98d7deac9f10102d7a8c8fd
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-profile-server

---
This message was auto-generated by https://gitcop.com

**Comment by GitCop** [comment]
Thanks for contributing! Unfortunately, I'm here to tell you there were the following style issues with your Pull Request:

 * Commit: 9053c7baaf7e1b9e7ba5ddf1df65a39d0c75511c
  * Invalid type. Valid types are **feat, fix, docs, style, refactor, perf, test, chore, revert**

Guidelines are available at https://github.com/mozilla/fxa-profile-server

---
This message was auto-generated by https://gitcop.com

**Comment by vladikoff** [comment]
@larissagaulia nice!

**Pull request merged**